src/mscorlib/shared/System/Buffers/ConfigurableArrayPool.cs

```diff
@@ -45,7 +46,11 @@ internal ConfigurableArrayPool(int maxArrayLength, int maxArraysPerBucket)
 
             // Create the buckets.
             int poolId = Id;
+#if !CORERT && !ARM && !ARM64
```

Comment:
Better way to do this? No point to test on ARM and isn't helpful for AoT as it won't inline and eliminate the other branch

Reply:
Why can't you just do `Utilities.SelectBucketIndex`, and have the `Lzcnt.IsSupported` check in there?

The checks are meant to be constant and the irrelevant code path should be completely optimizated away (non-intrinsic gets dropped if `IsSupported` is true, intrinsic path gets dropped if `IsSupported` is false, such as on ARM).

Reply:
This kind of shows that the current approach to intrinsics isn't that great in some cases. There really should be a platform independent "lzcnt" function somewhere that does the right thing on every platform.

Reply:
That is:

```
[MethodImpl(MethodImplOptions.AggressiveInlining)]
internal static int GetMaxSizeForBucket(int binIndex)
{
	if (LzCnt.IsSupported)
	{
		return (31 - (int)Lzcnt.LeadingZeroCount((uint)(binIndex - 1) >> 3)); 
	}
	else
	{
		int maxSize = 16 << binIndex;
		Debug.Assert(maxSize >= 0);
		return maxSize;
	}
}
```

and if CoreRT is a concern (due to not doing constant folding on `IsSupported` yet), change it to:

```
[MethodImpl(MethodImplOptions.AggressiveInlining)]
internal static int GetMaxSizeForBucket(int binIndex)
{
#if !CORERT
	if (LzCnt.IsSupported)
	{
		return (31 - (int)Lzcnt.LeadingZeroCount((uint)(binIndex - 1) >> 3)); 
	}
	else
#endif
	{
		int maxSize = 16 << binIndex;
		Debug.Assert(maxSize >= 0);
		return maxSize;
	}
}
```

Reply:
Only for Jitted code not for crossgened code (as System.Private.CoreLib is)

Testing `Lzcnt.IsSupported` in `SelectBucketIndex` causes the function to not crossgen; which means it no longer force inlines into `Rent`/`Return` (which do get crossgened) and the extra method call will outweigh any gain from `Lzcnt`

Moving the test into `Rent`/`Return` causes them to be non-crossgen'd; so the get compiled at Jit time when the branches get elminated and the `SelectBucketIndex{Intrinsic}` method inlined.

jit-diff shows `Rent`/`Return` drop out of the AoT image:
```
Total bytes of diff: -10767 (-0.32% of base)
    diff is an improvement.

Total byte diff includes -8897 bytes from reconciling methods
        Base had    5 unique methods,     8897 unique bytes
        Diff had    0 unique methods,        0 unique bytes

Top file improvements by size (bytes):
      -10767 : System.Private.CoreLib.dasm (-0.32% of base)

1 total files with size differences (1 improved, 0 regressed), 0 unchanged.

Top method improvements by size (bytes):
       -2788 : System.Private.CoreLib.dasm - TlsOverPerCoreLockedStacksArrayPool`1:Rent(int):ref:this (3/0 methods)
       -2191 : System.Private.CoreLib.dasm - TlsOverPerCoreLockedStacksArrayPool`1:Return(ref,bool):this (3/0 methods)
       -1849 : System.Private.CoreLib.dasm - ConfigurableArrayPool`1:Rent(int):ref:this (3/0 methods)
       -1163 : System.Private.CoreLib.dasm - ConfigurableArrayPool`1:.ctor(int,int):this (3/0 methods)
        -938 : System.Private.CoreLib.dasm - Bucket:Rent():ref:this (3/1 methods)

8 total methods with size differences (8 improved, 0 regressed), 16760 unchanged.
```

Reply:
Ah, ok. This is due to all the non-jit places not understanding intrinsics yet.

Reply:
Issue is the function won't inline with crossgen so it adds the cost of a method call when its not supported; then when it is supported the gain needs to be more than the additional method call cost.

Reply:
> Ah, ok. This is due to all the non-jit places not understanding intrinsics yet.

If it did inline it would have a 60 byte conditional jump in the crossgen it so it still wouldn't be quite as good

Reply:
Yeah, this is part of the CrossGen/AOT issue to determine how intrinsics should be compiled (dynamic dispatch, always JIT, etc).

The eventual logic probably needs to take these type of scenarios into account so that `SelectBucketIndex` can contain all the intrinsic logic.